### ISSUE TITLE
Intermittent NPE in dependency resolution after upgrading to 8.3

### ISSUE THREAD

**Issue by jeantil** [gradle/gradle#26298]
### Current Behavior

When upgrading the monorepo I work on from gradle 8.2.1 to 8.3, I encounter what appears to be semi random `NullPointerExceptions` during dependency resolution. 

Gradle displays the following output
```
Could not determine the dependencies of task ':business-referral-backend:bootJar'.
> Could not resolve all dependencies for configuration ':business-referral-backend:runtimeClasspath'.
   > java.lang.NullPointerException (no error message)
```
or
```
Could not determine the dependencies of task ':account-settings-backend:compileTestJava'.
> Could not resolve all dependencies for configuration ':account-settings-backend:testAnnotationProcessor'.
   > Cannot invoke "org.gradle.api.provider.Provider.get()" because the return value of "java.util.Map.get(Object)" is null
```
or 
```
Could not determine the dependencies of task ':account-settings-backend:compileTestJava'.
> Could not resolve all dependencies for configuration ':account-settings-backend:testAnnotationProcessor'.
   > Cannot invoke "org.gradle.api.provider.Provider.get()" because the return value of "java.util.Map.get(Object)" is null
```

The no error message in the first output is because of repeat exception erasure by java 17. 

### Expected Behavior

No `NullPointerException` :)

### Context (optional)

The root cause has proven very elusive so far : 
- some colleagues never have the errors, some have it consistently
- restarting the gradle daemon seems to sometimes fix the error I couldn't establish if it would fix it consistently
- enabling/disabling config-on-demand desn't change the status (error or no error) 
- CI builds with `--no-configure-on-demand --no-daemon` exhibit the issue
- reverting to 8.2.1 always fixed the NPE
- In IJ running a task which exhibits the failure does fail, reruning the same task in debug doesn't.

A sample full stack trace : 
```
org.gradle.api.internal.tasks.TaskDependencyResolveException: Could not determine the dependencies of task ':account-settings-backend:compileTestJava'.
	at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext.getDependencies(CachingTaskDependencyResolveContext.java:68)
	at org.gradle.execution.plan.TaskDependencyResolver.resolveDependenciesFor(TaskDependencyResolver.java:49)
	at org.gradle.execution.plan.LocalTaskNode.getDependencies(LocalTaskNode.java:148)
	at org.gradle.execution.plan.LocalTaskNode.resolveDependencies(LocalTaskNode.java:122)
	at org.gradle.execution.plan.DefaultExecutionPlan.discoverNodeRelationships(DefaultExecutionPlan.java:184)
	at org.gradle.execution.plan.DefaultExecutionPlan.doAddEntryNodes(DefaultExecutionPlan.java:156)
	at org.gradle.execution.plan.DefaultExecutionPlan.addEntryTasks(DefaultExecutionPlan.java:131)
	at org.gradle.execution.plan.DefaultExecutionPlan.addEntryTasks(DefaultExecutionPlan.java:123)
	at org.gradle.execution.TaskNameResolvingBuildTaskScheduler.scheduleRequestedTasks(TaskNameResolvingBuildTaskScheduler.java:54)
	at org.gradle.execution.DefaultTasksBuildTaskScheduler.scheduleRequestedTasks(DefaultTasksBuildTaskScheduler.java:72)
	at org.gradle.initialization.DefaultTaskExecutionPreparer.lambda$scheduleRequestedTasks$0(DefaultTaskExecutionPreparer.java:46)
	at org.gradle.internal.Factories$1.create(Factories.java:31)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withReplacedLocks(DefaultWorkerLeaseService.java:360)
	at org.gradle.api.internal.project.DefaultProjectStateRegistry$DefaultBuildProjectRegistry.withMutableStateOfAllProjects(DefaultProjectStateRegistry.java:228)
	at org.gradle.api.internal.project.DefaultProjectStateRegistry$DefaultBuildProjectRegistry.withMutableStateOfAllProjects(DefaultProjectStateRegistry.java:221)
	at org.gradle.initialization.DefaultTaskExecutionPreparer.scheduleRequestedTasks(DefaultTaskExecutionPreparer.java:45)
	at org.gradle.initialization.VintageBuildModelController.lambda$scheduleRequestedTasks$0(VintageBuildModelController.java:76)
	at org.gradle.internal.model.StateTransitionController.lambda$inState$1(StateTransitionController.java:99)
	at org.gradle.internal.model.StateTransitionController.lambda$inState$2(StateTransitionController.java:114)
	at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:44)
	at org.gradle.internal.model.StateTransitionController.inState(StateTransitionController.java:110)
	at org.gradle.internal.model.StateTransitionController.inState(StateTransitionController.java:98)
	at org.gradle.initialization.VintageBuildModelController.scheduleRequestedTasks(VintageBuildModelController.java:76)
	at org.gradle.internal.build.DefaultBuildLifecycleController$DefaultWorkGraphBuilder.addRequestedTasks(DefaultBuildLifecycleController.java:351)
	at org.gradle.internal.buildtree.DefaultBuildTreeWorkPreparer.lambda$scheduleRequestedTasks$0(DefaultBuildTreeWorkPreparer.java:41)
	at org.gradle.internal.build.DefaultBuildLifecycleController.lambda$populateWorkGraph$7(DefaultBuildLifecycleController.java:186)
	at org.gradle.internal.build.DefaultBuildWorkPreparer.populateWorkGraph(DefaultBuildWorkPreparer.java:42)
	at org.gradle.internal.build.BuildOperationFiringBuildWorkPreparer$PopulateWorkGraph.populateTaskGraph(BuildOperationFiringBuildWorkPreparer.java:106)
	at org.gradle.internal.build.BuildOperationFiringBuildWorkPreparer$PopulateWorkGraph.run(BuildOperationFiringBuildWorkPreparer.java:92)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
	at org.gradle.internal.build.BuildOperationFiringBuildWorkPreparer.populateWorkGraph(BuildOperationFiringBuildWorkPreparer.java:67)
	at org.gradle.internal.build.DefaultBuildLifecycleController.lambda$populateWorkGraph$8(DefaultBuildLifecycleController.java:186)
	at org.gradle.internal.model.StateTransitionController.lambda$inState$1(StateTransitionController.java:99)
	at org.gradle.internal.model.StateTransitionController.lambda$inState$2(StateTransitionController.java:114)
	at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:44)
	at org.gradle.internal.model.StateTransitionController.inState(StateTransitionController.java:110)
	at org.gradle.internal.model.StateTransitionController.inState(StateTransitionController.java:98)
	at org.gradle.internal.build.DefaultBuildLifecycleController.populateWorkGraph(DefaultBuildLifecycleController.java:186)
	at org.gradle.internal.build.DefaultBuildWorkGraphController$DefaultBuildWorkGraph.populateWorkGraph(DefaultBuildWorkGraphController.java:169)
	at org.gradle.composite.internal.DefaultBuildController.populateWorkGraph(DefaultBuildController.java:76)
	at org.gradle.composite.internal.DefaultIncludedBuildTaskGraph$DefaultBuildTreeWorkGraphBuilder.withWorkGraph(DefaultIncludedBuildTaskGraph.java:153)
	at org.gradle.internal.buildtree.DefaultBuildTreeWorkPreparer.lambda$scheduleRequestedTasks$1(DefaultBuildTreeWorkPreparer.java:41)
	at org.gradle.composite.internal.DefaultIncludedBuildTaskGraph$DefaultBuildTreeWorkGraph$1.run(DefaultIncludedBuildTaskGraph.java:209)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
	at org.gradle.composite.internal.DefaultIncludedBuildTaskGraph$DefaultBuildTreeWorkGraph.scheduleWork(DefaultIncludedBuildTaskGraph.java:204)
	at org.gradle.internal.buildtree.DefaultBuildTreeWorkPreparer.scheduleRequestedTasks(DefaultBuildTreeWorkPreparer.java:37)
	at org.gradle.configurationcache.VintageBuildTreeWorkController$scheduleAndRunRequestedTasks$1.apply(VintageBuildTreeWorkController.kt:36)
	at org.gradle.configurationcache.VintageBuildTreeWorkController$scheduleAndRunRequestedTasks$1.apply(VintageBuildTreeWorkController.kt:35)
	at org.gradle.composite.internal.DefaultIncludedBuildTaskGraph.withNewWorkGraph(DefaultIncludedBuildTaskGraph.java:112)
	at org.gradle.configurationcache.VintageBuildTreeWorkController.scheduleAndRunRequestedTasks(VintageBuildTreeWorkController.kt:35)
	at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$scheduleAndRunTasks$1(DefaultBuildTreeLifecycleController.java:68)
	at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$runBuild$4(DefaultBuildTreeLifecycleController.java:98)
	at org.gradle.internal.model.StateTransitionController.lambda$transition$6(StateTransitionController.java:169)
	at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:266)
	at org.gradle.internal.model.StateTransitionController.lambda$transition$7(StateTransitionController.java:169)
	at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:44)
	at org.gradle.internal.model.StateTransitionController.transition(StateTransitionController.java:169)
	at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.runBuild(DefaultBuildTreeLifecycleController.java:95)
	at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.scheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:68)
	at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.scheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:63)
	at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:31)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.internal.buildtree.ProblemReportingBuildActionRunner.run(ProblemReportingBuildActionRunner.java:49)
	at org.gradle.launcher.exec.BuildOutcomeReportingBuildActionRunner.run(BuildOutcomeReportingBuildActionRunner.java:65)
	at org.gradle.tooling.internal.provider.FileSystemWatchingBuildActionRunner.run(FileSystemWatchingBuildActionRunner.java:140)
	at org.gradle.launcher.exec.BuildCompletionNotifyingBuildActionRunner.run(BuildCompletionNotifyingBuildActionRunner.java:41)
	at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.lambda$execute$0(RootBuildLifecycleBuildActionExecutor.java:40)
	at org.gradle.composite.internal.DefaultRootBuildState.run(DefaultRootBuildState.java:123)
	at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.execute(RootBuildLifecycleBuildActionExecutor.java:40)
	at org.gradle.internal.buildtree.InitDeprecationLoggingActionExecutor.execute(InitDeprecationLoggingActionExecutor.java:58)
	at org.gradle.internal.buildtree.DefaultBuildTreeContext.execute(DefaultBuildTreeContext.java:40)
	at org.gradle.launcher.exec.BuildTreeLifecycleBuildActionExecutor.lambda$execute$0(BuildTreeLifecycleBuildActionExecutor.java:65)
	at org.gradle.internal.buildtree.BuildTreeState.run(BuildTreeState.java:58)
	at org.gradle.launcher.exec.BuildTreeLifecycleBuildActionExecutor.execute(BuildTreeLifecycleBuildActionExecutor.java:65)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$3.call(RunAsBuildOperationBuildActionExecutor.java:61)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$3.call(RunAsBuildOperationBuildActionExecutor.java:57)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor.execute(RunAsBuildOperationBuildActionExecutor.java:57)
	at org.gradle.launcher.exec.RunAsWorkerThreadBuildActionExecutor.lambda$execute$0(RunAsWorkerThreadBuildActionExecutor.java:36)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:264)
	at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:128)
	at org.gradle.launcher.exec.RunAsWorkerThreadBuildActionExecutor.execute(RunAsWorkerThreadBuildActionExecutor.java:36)
	at org.gradle.tooling.internal.provider.continuous.ContinuousBuildActionExecutor.execute(ContinuousBuildActionExecutor.java:110)
	at org.gradle.tooling.internal.provider.SubscribableBuildActionExecutor.execute(SubscribableBuildActionExecutor.java:64)
	at org.gradle.internal.session.DefaultBuildSessionContext.execute(DefaultBuildSessionContext.java:46)
	at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter$ActionImpl.apply(BuildSessionLifecycleBuildActionExecuter.java:92)
	at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter$ActionImpl.apply(BuildSessionLifecycleBuildActionExecuter.java:80)
	at org.gradle.internal.session.BuildSessionState.run(BuildSessionState.java:69)
	at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter.execute(BuildSessionLifecycleBuildActionExecuter.java:62)
	at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter.execute(BuildSessionLifecycleBuildActionExecuter.java:41)
	at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:64)
	at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:32)
	at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:51)
	at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:39)
	at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:47)
	at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:31)
	at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:65)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:39)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:29)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:35)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:78)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:75)
	at org.gradle.util.internal.Swapper.swap(Swapper.java:38)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:75)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:53)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:63)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:84)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:52)
	at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:297)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
Caused by: org.gradle.api.artifacts.ResolveException: Could not resolve all dependencies for configuration ':account-settings-backend:testAnnotationProcessor'.
	at org.gradle.api.internal.artifacts.ResolveExceptionContextualizer.contextualize(ResolveExceptionContextualizer.java:55)
	at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver.resolveBuildDependencies(ErrorHandlingConfigurationResolver.java:65)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.lambda$resolveGraphForBuildDependenciesIfRequired$5(DefaultConfiguration.java:912)
	at org.gradle.api.internal.project.DefaultProjectStateRegistry$CalculatedModelValueImpl.update(DefaultProjectStateRegistry.java:493)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolveGraphForBuildDependenciesIfRequired(DefaultConfiguration.java:909)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.access$1800(DefaultConfiguration.java:173)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$VisitedArtifactsSetProvider.getTaskDependencyValue(DefaultConfiguration.java:1656)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$VisitedArtifactsSetProvider.getTaskDependencyValue(DefaultConfiguration.java:1651)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$SelectedArtifactsProvider.getTaskDependencyValue(DefaultConfiguration.java:1693)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$SelectedArtifactsProvider.getTaskDependencyValue(DefaultConfiguration.java:1667)
	at org.gradle.api.internal.artifacts.configurations.ResolutionBackedFileCollection.visitDependencies(ResolutionBackedFileCollection.java:56)
	at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext$TaskGraphImpl.getNodeValues(CachingTaskDependencyResolveContext.java:103)
	at org.gradle.internal.graph.CachingDirectedGraphWalker$GraphWithEmptyEdges.getNodeValues(CachingDirectedGraphWalker.java:213)
	at org.gradle.internal.graph.CachingDirectedGraphWalker.doSearch(CachingDirectedGraphWalker.java:121)
	at org.gradle.internal.graph.CachingDirectedGraphWalker.findValues(CachingDirectedGraphWalker.java:73)
	at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext.getDependencies(CachingTaskDependencyResolveContext.java:66)
	... 142 more
Caused by: java.lang.NullPointerException: Cannot invoke "org.gradle.api.provider.Pro
vider.get()" because the return value of "java.util.Map.get(Object)" is null
	at org.gradle.api.internal.attributes.DefaultMutableAttributeContainer.lambda$realizeAllLazyAttributes$4(DefaultMutableAttributeContainer.java:194)
	at org.gradle.api.internal.attributes.DefaultMutableAttributeContainer.realizeAllLazyAttributes(DefaultMutableAttributeContainer.java:194)
	at org.gradle.api.internal.attributes.DefaultMutableAttributeContainer.asImmutable(DefaultMutableAttributeContainer.java:139)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.preventFromFurtherMutation(DefaultConfiguration.java:1153)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.preventFromFurtherMutation(DefaultConfiguration.java:1142)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfigurationContainer.visitAll(DefaultConfigurationContainer.java:[82](https://gitlab.com/maltcommunity/malt/apps/malt/-/jobs/5027737641#L82))
	at org.gradle.api.internal.artifacts.ivyservice.moduleconverter.DefaultRootComponentMetadataBuilder.getRootComponentMetadata(DefaultRootComponentMetadataBuilder.java:111)
	at org.gradle.api.internal.artifacts.ivyservice.moduleconverter.DefaultRootComponentMetadataBuilder.lambda$buildRootComponentMetadata$0(DefaultRootComponentMetadataBuilder.java:99)
	at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.fromMutableState(DefaultProjectStateRegistry.java:413)
	at org.gradle.api.internal.artifacts.ivyservice.moduleconverter.DefaultRootComponentMetadataBuilder.buildRootComponentMetadata(DefaultRootComponentMetadataBuilder.java:97)
	at org.gradle.api.internal.artifacts.ivyservice.moduleconverter.DefaultRootComponentMetadataBuilder.toRootComponentMetaData(DefaultRootComponentMetadataBuilder.java:[83](https://gitlab.com/maltcommunity/malt/apps/malt/-/jobs/5027737641#L83))
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.toRootComponentMetaData(DefaultConfiguration.java:[137](https://gitlab.com/maltcommunity/malt/apps/malt/-/jobs/5027737641#L137)9)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.builder.DependencyGraphBuilder.resolve(DependencyGraphBuilder.java:[138](https://gitlab.com/maltcommunity/malt/apps/malt/-/jobs/5027737641#L138))
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.DefaultArtifactDependencyResolver.resolve(DefaultArtifactDependencyResolver.java:157)
	at org.gradle.api.internal.artifacts.ivyservice.DefaultConfigurationResolver.resolveBuildDependencies(DefaultConfigurationResolver.java:161)
	at org.gradle.api.internal.artifacts.ivyservice.ShortCircuitEmptyConfigurationResolver.resolveBuildDependencies(ShortCircuitEmptyConfigurationResolver.java:77)
	at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver.resolveBuildDependencies(ErrorHandlingConfigurationResolver.java:63)
	... [156](https://gitlab.com/maltcommunity/malt/apps/malt/-/jobs/5027737641#L156) more
```

### Steps to Reproduce

As explained in context, I have yet to find a way to consistently reproduce the error or to consistently fix it.  

I raised the issue on the [gradle community slack](https://gradle-community.slack.com/archives/CAHSN3LDN/p1694002410874909) and Vampire told me to open an issue 

>Oh, well, great code, but it is in there since 7.5.0 already: https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java#L191-L195
No wonder it fails sometimes.
To prevent a concurrent modification exception, they get the keys of a map, then iterate over the keys and get the value from the map, calling get() on the result.
Well, actually it is also not a thread-safe map, but a simple LinkedHashMap.
I'd suppose it is used from different threads and even if it were a thread-safe map, it could be modified by another thread while iterating over the keys.
I'd say you should definitely open a bug report.
No idea why it behaves differently in 8.3, maybe the code that modifies the map was introduce there or something like that.

### Gradle version

8.3

### Build scan URL (optional)

https://scans.gradle.com/s/6f36pgrqzwnz2/failure?expanded-stacktrace=WyIwIl0#1

### Your Environment (optional)

I have reports of this error using gradle 8.3 on individual laptops running either ubuntu or mac but all on java 17, I have errors on CI servers also with java 17.

**Comment by big-guy** [comment]
This is very suspicious looking. 

Do you have a build scan for a successful run?  That might help create a reproducer with a similar looking dependency graph.

**Comment by jeantil** [comment]
> Do you have a build scan for a successful run? That might help create a reproducer with a similar looking dependency graph.
using gradle 8.2.1 or 8.3 ?

**Comment by big-guy** [comment]
Both if you can

**Comment by jeantil** [comment]
gradle 8.2.1 https://scans.gradle.com/s/vorh7l5x2c4ru
8.3 will take some time as upgrading invalidates a lot of caches ...

**Comment by jeantil** [comment]
8.3 https://scans.gradle.com/s/olyrirko67wbm this upgrade attempt didn't trigger the error ...

**Comment by big-guy** [comment]
@jeantil I haven't been able to reproduce the failure you're seeing yet.

What seems to be happening is something is concurrently realizing the same attribute container while Gradle is building the task graph and resolving a configuration.  

Building the task graph is single-threaded, so I'm trying to rule out some possibilities:
- There's a "Scan took" message in all the console logs.  https://scans.gradle.com/s/vorh7l5x2c4ru/console-log?page=1#L17 What is this scan referring to? 
- Do any of your local plugins (`com.malt.*`) do things like: resolve all configurations in a project, spawn a background thread to resolve a configuration or look at the Project state, use the Gradle Enterprise plugin's `background {}` hook.

Would you be able to try out a nightly if we needed to collect some additional logging/information from your build in particular?

With Gradle 8.3, could you put this in the bottom of your crm-app build file:
```
configurations {
    runtimeClasspath {
       attributes {
           attributeProvider(Attribute.of("ignored", Named::class.java), provider {
               val threadName = Thread.currentThread().name
               System.out.println(threadName)
               Throwable().printStackTrace()
               objects.named(Named::class.java, "ignored")
           })
       }
   }
}
```
and try to reproduce the failure by running`crm-app:bootJar`?

This will print some extra logging that will hopefully point us towards who else is realizing this container.  Even if you don't see the failure with this logging, could you share a build scan with it?

**Comment by jeantil** [comment]
Thanks for looking into this :pray: 

1. About scan took, 

The build registers `excludeDirs` for intellij using the idea module mostly to exlude frontend related files.
```
idea {
    module {
        val start = Instant.now()
        excludeDirs.add(file("**redacted**"))
        excludeDirs.add(file("**redacted**"))
        excludeDirs.add(file("**redacted**"))
        excludeDirs.add(file("**redacted**"))
        excludeDirs.add(file("**redacted**"))
        excludeDirs.addAll(listExclusions(rootDir))
        val end = Instant.now()
        val duration = Duration.between(start, end)

        logger.lifecycle("Scan took $duration")
    }
}

fun listExclusions(parent: File): Set<File> {
    logger.info("Entering ${parent.path}")

    val files = if (parent.isDirectory) parent.listFiles()!! else return emptySet()

    val foldersToExclude = mutableSetOf<File>()

    if(parentIsRootDir(parent) || parentIsItsOwnGradleModule(files)) {
        files.filterNot { scanExcludedDirs.contains(it.name) }
            .forEach {
                if (it.isDirectory) {
                    foldersToExclude.addAll(listExclusions(it))
                } else if (it.name.equals("package.json")) {
                    foldersToExclude.addAll(frontFoldersToExclude(it.path))
                }
            }
    }

    return foldersToExclude
}
```
>  about local plugins

  * resolve all configurations in a project, 
    => not that I am aware of, if one does it is accidental
  * spawn a background thread to resolve a configuration or look at the Project state, 
    => same as the previous one
  * use the Gradle Enterprise plugin's background {} hook.
    =>  this one I can definitely rule out

> Would you be able to try out a nightly if we needed to collect some additional logging/information from your build in particular?

yes sure

> With Gradle 8.3, could you put this in the bottom of your crm-app build file: [... ] and try to reproduce the failure by runningcrm-app:bootJar?

I haven't managed to reproduce the failure :/ but I did create [a scan with the requested configuration](https://scans.gradle.com/s/foanikxcqftck).

**Comment by jeantil** [comment]
I tried a few things : 
killing the daemon reverting back to 8.2.1, dropping caches related to 8.3,then running a few tasks on the build 
then upgrading again but I can't reproduce the error ...

I also checked the history for buildSrc, the toplevel build.gradle.kts, settings.gradle.kts there was only a single line change which couldn't be related to the failure (enabled by property in specific dev envs only) 
crm-app's build file only registers added or removed  implementation dependencies to jvm artifacts nothing fancy that would explain that it did fail before and doesn't anymore now :(

**Comment by jeantil** [comment]
I spent part of the day trying to reproduce the issue again an this time for the exact same commit, I have a successful build scan and a failed one ! 

success : https://scans.gradle.com/s/foanikxcqftck
failure: https://scans.gradle.com/s/b6a2snl7jarcc

this is on the exact same codebase with flags as close as I could manage

**Comment by big-guy** [comment]
Awesome, thanks.

In the last hour, I was able to reproduce the exception with a small reproducer.  Your logs make me think it's the same underlying cause. Thankfully, it's not something concurrency/multi-thread related.  I also found where I think the non-deterministic behavior is coming from. 

Hopefully I'll have a nightly for you to try soon (tonight/tomorrow).

**Comment by big-guy** [comment]
Could you give this nightly a try?  You can remove the hacky thing I mentioned above. 

```
gradle wrapper --gradle-version=8.4-branch-sg_84_regression_attribute_ordering-20230907211623+0000
```

This is based on #26315 

My expectation is that this should fix the NPE.  I'm currently running our test suite to see if there are any surprises.

**Comment by jeantil** [comment]
Hello, I have tried the nightly this morning. 
It's hard to tell if it's really fixed because of the intermittent nature but it is looking good
I have run 
``` 
./gradlew  wrapper --gradle-version=8.3
./gradlew assemble -Dfile.encoding=UTF-8 -Dkotlin.environment.keepalive=true --no-configure-on-demand  --stacktrace
./gradlew  wrapper --gradle-version=8.4-branch-sg_84_regression_attribute_ordering-20230907211623+0000
./gradlew assemble -Dfile.encoding=UTF-8 -Dkotlin.environment.keepalive=true --no-configure-on-demand  --stacktrace
```
 a few times and never triggered the NPE with 8.4 while I did trigger it with 8.3 including when downgrading from 8.4
I pushed the 8.4 snapshot version to CI and didn't trigger the error
:crossed_fingers:

**Comment by Vampire** [comment]
From the code changes, I'd also say the NPE should not be able to occur again. :-)
👌

**Comment by jeantil** [comment]
do you think this would be included in an 8.3.1 or will we have to wait for 8.4 ?

**Comment by SgtSilvio** [comment]
Running into the same issue (https://github.com/hivemq/mqtt-cli/actions/runs/6160887092/job/16723111964).
Can confirm that it works with 8.4-rc-1.
